### PR TITLE
Update deps and support knex v3

### PR DESCRIPTION
**Pull request by denwilliams**
Allow the project to work with knex v3 as well as v2

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -4,17 +4,38 @@ on:
...
-      - uses: actions/setup-node@v3
+      - uses: actions/checkout@v4
+      - uses: actions/setup-node@v4
+        with:
+          node-version: 20
+      - run: npm ci
+      - run: npm i -D prom-client@${{matrix.prom-client-version}}
+      - run: npm i -D knex@${{matrix.knex-version}}
+      - run: npm run build
+  build_knex2: # knex 2 won't compile with typescript 5
```
Comment (denwilliams) [inline .github/workflows/build.yml:23]: knex 2 doesnt work with typescript v5, and typescript 4 doesnt work with other things

so we need to install some older versions of typings that work with typescript 4

we should remove knex v2 support in the near future

File: knex/index.ts [hunk 2]
```diff
@@ -222,7 +223,7 @@ export async function setup(
   }
 }
 
-export function registerMetrics(registry: Registry) {
+export function registerMetrics(registry: Pick<Registry, "registerMetric">) {
```
Comment (denwilliams) [inline knex/index.ts:226]: This just stops some weird version compat things happening

Comment (smithamax) [reply knex/index.ts:226]: Alex and I did the same thing with redis https://github.com/LOKE/loke-redis-cache/blob/master/index.ts#L314

File: .github/workflows/build.yml (hunk 1 above)
Comment (smithamax) [reply .github/workflows/build.yml:23]: In the redis lib we do a similar thing, but it works locally too. It was a bit sketchy to get working, and this way you're it doing is more flexible

https://github.com/LOKE/loke-redis-cache/blob/master/package.json#L33-L34

**Review by smithamax — APPROVED** [review]
Would be nice to get some basic sanity check tests in here. Happy to go ahead without if this is blocking something else.

If the multi version thing is brittle we can always just major version bump and make it only support knex 3

**Comment by denwilliams** [comment]
Yeah it was blocking me from updating knex, which was blocking something else, which was blocking updating node 20

**Pull request merged**